### PR TITLE
feature/openid_improvements

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
===========
OpenID Improvements

Impacts
=======
OpenID workflow

NEWS file entries
=======================

Enhancements
-------------
* OpenID pid mapping is now configurable
* Can map OpenID attributes to a person attributes
* Allow to create authentication rules based on OpenID attributes

Delete branch after merge
=========================
NO

Checklist
=========
- [ ] Document the feature
- [x] Add unit tests
- [ ] Add acceptance tests (TestLink)

### CODE REVIEW

**Comment by extrafu** [comment]
@satkunas Please do the require GUI changes on that PR.

**Comment by satkunas** [comment]
@jrouzierinverse please outline the needed changes w/ the front-end...

**Comment by jrouzierinverse** [comment]
* New parameter in pf.conf advanced.openid_attributes

* Update the OpenID Parameters  ```/usr/local/pf/sbin/pfperl-api get -M OPTIONS /api/v1/config/sources?type=OpenID | jq .meta.person_mappings```

**Comment by satkunas** [comment]
@jrouzierinverse front end pushed. Should be working, but I still haven't tested it with a real config.

**Comment by satkunas** [comment]
More work needed on group errors

File: html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/OAuth/OpenID.pm [hunk 1]
```diff
@@ -17,6 +17,29 @@ has '+source' => (isa => 'pf::Authentication::Source::OpenIDSource');
...
+    if (!exists $info->{$username}) {
+        return $self->SUPER::_extract_username_from_response($info);
+    }
+
+    return $info->{$source->username_attribute} // $self->SUPER::_extract_username_from_response($info);
+}
+
+sub auth_source_params_child {
+    my ($self) = @_;
+    my $info = person_view($self->username());
```
Comment (nqb) [inline html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/OAuth/OpenID.pm:40]: I got following error message when testing basic OpenID workflow:
```
Caught exception in captiveportal::Controller::Root->dynamic_application
"Undefined subroutine
&captiveportal::PacketFence::DynamicRouting::Module::Authentication::OAuth::OpenID::person_view
called at
/usr/local/pf/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/OAuth/OpenID.pm
line 39."
```
but `pf::person` is used here: https://github.com/inverse-inc/packetfence/blob/a60ff1f00e2a929377ffaeed39f27910909aa87d/html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication.pm#L20

Comment (nqb) [reply html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/OAuth/OpenID.pm:40]: if I use `use pf::person` directly in `html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/OAuth/OpenID.pm`, it works.

File: lib/pf/Authentication/Source/OpenIDSource.pm [hunk 2]
```diff
@@ -33,6 +52,64 @@ Which module to use for DynamicRouting
...
+
+lookup_from_provider_info
+
+=cut
+
+sub lookup_from_provider_info {
+    my ($self, $pid, $info) = @_;
+    my $person_info = $self->map_to_person($info);
+    if ($person_info) {
+        person_modify($pid, %$person_info);
```
Comment (nqb) [inline lib/pf/Authentication/Source/OpenIDSource.pm:66]: I got following error when PF try to modify an existing user:
```
Caught exception in captiveportal::Controller::Root->dynamic_application "Undefined subroutine &pf::Authentication::Source::OpenIDSource::person_modify called at /usr/local/pf/lib/pf/Authentication/Source/OpenIDSource.pm line 65."
```

**Review by nqb — CHANGES_REQUESTED** [review]
@satkunas, I'm not able to see the JS code that add `advanced.openid_attributes` in GUI.

Did you missed it ?

File: html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/OAuth/OpenID.pm (hunk 1 above)
Comment (jrouzierinverse) [reply html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/OAuth/OpenID.pm:40]: I push a fix for that error.

File: lib/pf/Authentication/Source/OpenIDSource.pm (hunk 2 above)
Comment (jrouzierinverse) [reply lib/pf/Authentication/Source/OpenIDSource.pm:66]: Fixed the import

**Comment by nqb** [comment]
With following config:

### `pf.conf`
```
# grep openid pf.conf
openid_attributes=email,birthdate
```

### `authentication.conf`
```
[yahoo]
create_local_account=no
client_secret=REDACTED
password_length=8
access_token_path=https://api.login.yahoo.com/oauth2/get_token
authorize_path=https://api.login.yahoo.com/oauth2/request_auth
set_access_durations_action=
protected_resource_url=https://api.login.yahoo.com/openid/v1/userinfo
scope=openid
local_account_logins=0
client_id=REDACTED
description=yahoo openid
domains=*
hash_passwords=bcrypt
site=https://api.login.yahoo.com
redirect_url=https://pfcen7dev.lan/oauth2/callback
type=OpenID
person_mapping.2=lang:locale
person_mapping.1=birthday:birthdate
person_mapping.0=email:email

[yahoo rule birthdate]
action0=set_role=guest
condition0=birthdate,contains,1989
status=enabled
match=all
class=authentication
action1=set_access_duration=1h

[yahoo rule catchall]
action0=set_role=myrole
status=enabled
match=all
class=authentication
action1=set_access_duration=1h
```
I'm able to get a `guest` role if I was born in 1989.

It works as it should.

**Review by nqb — CHANGES_REQUESTED** [review]
@jrouzierinverse,

With current state of PR, it doesn't seem possible to select which openid attribute to use as username.

It means that OpenID source will always use `email` attribute to create person in DB. However, my customer want to use a specific `username` or some provider may want to not return an email attribute.

**Comment by jrouzierinverse** [comment]
Add ```username_attribute=<attribute>``` to the config.
Sorry @satkunas I was not explicit in which fields I added to the OpenID form.
Can you also add username_attribute to the OpenID Form.

**Comment by satkunas** [comment]
@jrouzierinverse I don't see a `username_attribute` field in the meta. Is this the field we wish to add?

```/usr/local/pf/sbin/pfperl-api get -M OPTIONS "/api/v1/config/sources?type=OpenID"| jq '.meta.username_attribute'```


...or do we want to add `username` to the list of available mappings and validate the mappings so it contains at least a `username` mapping?

Screenshot_20200602_105827

**Comment by jrouzierinverse** [comment]
@satkunas Sorry I missed that I only added it to the source it pushed the new commit

**Comment by extrafu** [comment]
@nqb / @jrouzierinverse  / @satkunas We need to get that in today or it's out for the v10.1 release.

**Comment by nqb** [comment]
@satkunas, I still didn't see OpenID attributes in advanced section.

**Comment by nqb** [comment]
Latest changes added by @jrouzierinverse and @satkunas work as expected.

The only remaining thing is to have OpenID attributes field in Advanced section.

**Pull request merged**